crates/worker/src/partition/state_machine/command_handler.rs

```diff
@@ -623,218 +599,240 @@ where
             inbox_sequence_number,
-        );
-        effects.free_invocation(invocation_id);
+        )
+        .await?;
+        Self::do_free_invocation(ctx, invocation_id).await?;
 
         self.notify_invocation_result(
+            ctx,
             invocation_id,
             invocation_target,
             span_context,
             MillisSinceEpoch::now(),
             Err((error.code(), error.to_string())),
-            effects,
         );
 
         Ok(())
     }
 
-    async fn kill_invocation<State: StateReader>(
+    async fn kill_invocation<State: StateReader + StateStorage + ReadOnlyInvocationStatusTable>(
         &mut self,
+        ctx: &mut StateMachineApplyContext<'_, State>,
         invocation_id: InvocationId,
         metadata: InFlightInvocationMetadata,
-        state: &mut State,
-        effects: &mut Effects,
     ) -> Result<(), Error> {
-        self.kill_child_invocations(
-            &invocation_id,
-            state,
-            effects,
-            metadata.journal_metadata.length,
-        )
-        .await?;
+        self.kill_child_invocations(ctx, &invocation_id, metadata.journal_metadata.length)
+            .await?;
 
-        self.fail_invocation(effects, invocation_id, metadata, KILLED_INVOCATION_ERROR)
+        self.fail_invocation(ctx, invocation_id, metadata, KILLED_INVOCATION_ERROR)
             .await?;
-        effects.abort_invocation(invocation_id);
+        Self::do_send_abort_invocation_to_invoker(ctx, invocation_id);
         Ok(())
     }
 
-    async fn kill_child_invocations<State: StateReader>(
+    async fn kill_child_invocations<State: StateReader + StateStorage>(
         &mut self,
+        ctx: &mut StateMachineApplyContext<'_, State>,
         invocation_id: &InvocationId,
-        state: &mut State,
-        effects: &mut Effects,
         journal_length: EntryIndex,
     ) -> Result<(), Error> {
-        let mut journal_entries = pin!(state.get_journal(invocation_id, journal_length));
-        while let Some(journal_entry) = journal_entries.next().await {
-            let (_, journal_entry) = journal_entry?;
-
-            if let JournalEntry::Entry(enriched_entry) = journal_entry {
-                let (h, _) = enriched_entry.into_inner();
-                match h {
-                    // we only need to kill child invocations if they are not completed and the target was resolved
-                    EnrichedEntryHeader::Call {
-                        is_completed,
-                        enrichment_result: Some(enrichment_result),
-                    } if !is_completed => {
-                        self.handle_outgoing_message(
-                            OutboxMessage::InvocationTermination(InvocationTermination::kill(
-                                enrichment_result.invocation_id,
-                            )),
-                            effects,
-                        );
+        let invocation_ids_to_kill: Result<Vec<InvocationId>, _> = ctx
```

Comment:
I guess you could directly propagate the error here by using the `?` after the `await`.